File: OBI2/poder.py
import sys
sys.setrecursionlimit(200000)

def dfs(i, j, matrix, max_power, visited, N, M):
    if visited[i][j]:
        return max_power[i][j]

    visited[i][j] = True
    power = matrix[i][j]
    max_p = power

    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]  # Up, Down, Left, Right
    for di, dj in directions:
        ni, nj = i + di, j + dj
        if 0 <= ni < N and 0 <= nj < M and power >= matrix[ni][nj]:
            max_p = max(max_p, power + dfs(ni, nj, matrix, max_power, visited, N, M))

    max_power[i][j] = max_p
    return max_p

def main():
    N, M = map(int, input().split())
    matrix = [list(map(int, input().split())) for _ in range(N)]
    
    max_power = [[0] * M for _ in range(N)]
    visited = [[False] * M for _ in range(N)]
    
    for i in range(N):
        for j in range(M):
            if not visited[i][j]:
                dfs(i, j, matrix, max_power, visited, N, M)
    
    for row in max_power:
        print(" ".join(map(str, row)))

if __name__ == "__main__":
    main()

File: OBI2024/OBI2/poder.java
import java.util.Scanner;

public class poder {
    private static int[][] matrix;
    private static int[][] maxPower;
    private static boolean[][] visited;
    private static int N, M;

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        N = in.nextInt();
        M = in.nextInt();
        matrix = new int[N][M];
        maxPower = new int[N][M];
        visited = new boolean[N][M];

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                matrix[i][j] = in.nextInt();
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (!visited[i][j]) {
                    dfs(i, j);
                }
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                System.out.print(maxPower[i][j] + " ");
            }
            System.out.println();
        }
    }

    private static int dfs(int i, int j) {
        if (visited[i][j]) {
            return maxPower[i][j];
        }

        visited[i][j] = true;
        int power = matrix[i][j];
        int max = power;

        // Move up
        if (i > 0 && power >= matrix[i - 1][j]) {
            max = Math.max(max, power + dfs(i - 1, j));
        }
        // Move down
        if (i < N - 1 && power >= matrix[i + 1][j]) {
            max = Math.max(max, power + dfs(i + 1, j));
        }
        // Move left
        if (j > 0 && power >= matrix[i][j - 1]) {
            max = Math.max(max, power + dfs(i, j - 1));
        }
        // Move right
        if (j < M - 1 && power >= matrix[i][j + 1]) {
            max = Math.max(max, power + dfs(i, j + 1));
        }

        maxPower[i][j] = max;
        return max;
    }
}